cpp/include/cugraph/algorithms.hpp

```diff
@@ -1416,6 +1416,64 @@ random_walks(raft::handle_t const& handle,
              bool use_padding                                     = false,
              std::unique_ptr<sampling_params_t> sampling_strategy = nullptr);
 
+/**
+ * @brief generate sub-sampled graph in CSR format given input graph, list of vertices
+ * and sample size per vertex. The output graph consists of the given vertices
+ * with each vertex having at most `sample_size` neighbors from the original graph
+ *
+ * @tparam graph_t Type of input graph/view (typically, graph_view_t).
+ * @tparam index_t Type used to store indexing and sizes.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph Graph (view )object to sub-sample.
+ * @param ptr_d_start Device pointer to set of starting vertex indices for the sub-sampling.
+ * @param num_vertices = number(vertices) to use for sub-sampling.
+ * @param sampling_size = max number of neighbors per output vertex.
+ * @param sampling_type = the sampling type (algo R/algo L/etc.) used to produce outputs.
+ * @return std::tuple<rmm::device_uvector<typename graph_t::edge_type>,
+ *                    rmm::device_uvector<typename graph_t::vertex_type>>
+ * Tuple consisting of two arrays representing the offsets and indices of
+ * the sub-sampled graph.
+ */
+template <typename graph_t, typename index_t>
+std::tuple<rmm::device_uvector<typename graph_t::edge_type>,
+           rmm::device_uvector<typename graph_t::vertex_type>>
+neighbor_sampling_csr(raft::handle_t const& handle,
+                      graph_t const& graph,
+                      typename graph_t::vertex_type const* ptr_d_start,
+                      index_t num_vertices,
+                      index_t sampling_size,
```

Comment:
Should we define `index_t` as a template type? (as this is used only for `num_vertices` and `sampling size`, no significant memory saving by using a smaller type) Can't we just use `size_t`?

Reply:
Yes, I took this from the random walks interface. In reality, we could even restrict this to `int` / `unisgned long` since at no point in the foreseeable future will the number of vertices or sampling size be larger than `INT_MAX`.

But we can use `size_t` and the implementation will have asserts in place to ensure that this isn't larger than `INT_MAX`. The kernels themselves will use `int` since integer arithmetic is significant for these kernels and there is no reason to perform that with 64 bits when 32 will do perfectly fine.

Reply:
Yes, I prefer size_t in the public API to be future proof, but if there is non-negligible performance penalty, we can convert to int inside the kernel.

Reply:
makes sense, I'll change it accordingly